sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala

```diff
@@ -191,13 +194,27 @@ private[sql] class SharedState(
 }
 
 object SharedState extends Logging {
-  try {
-    URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory())
-  } catch {
-    case e: Error =>
-      logWarning("URL.setURLStreamHandlerFactory failed to set FsUrlStreamHandlerFactory")
+  private def setFsUrlStreamHandlerFactory(conf: SparkConf): Unit = {
+    if (conf.get(DEFAULT_URL_STREAM_HANDLER_FACTORY_ENABLED) && factory.isEmpty) {
+      factory.synchronized {
+        if (factory.isEmpty) {
+          try {
+            URL.setURLStreamHandlerFactory(defaultFactory)
+            factory = Some(defaultFactory)
+          } catch {
+            case NonFatal(e) =>
+              logWarning("URL.setURLStreamHandlerFactory failed to set " +
+                "FsUrlStreamHandlerFactory", e)
+          }
+        }
+      }
+    }
   }
 
+  @volatile private var factory: Option[FsUrlStreamHandlerFactory] = None
+
+  private lazy val defaultFactory = new FsUrlStreamHandlerFactory()
```

Comment:
Back to the same comment. You don't need this reference, nor the one above. You just need some AtomicBoolean flag to call this exactly one.

Reply:
Ah, that was your point. Got it :)